[jit] Mark values entering containers as wildcards

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #20670 [jit] make wildcards alias only each other
* **#20556 [jit] Mark values entering containers as wildcards**

This is the first step to improving wildcard analysis.

The end-state is that the wildcard set now represents everything we're
not sure about. Inputs and values that enter/exit containers will be
part of the wildcard set. There will be one wildcard set per type.

The key difference is that today, a wildcard value may alias *any other
value* in the graph. After this change, the wildcard value may alias
*any other wildcard*, but will not alias other stuff.

So if we have a lot of purely functional code in an inner block, we can
optimize it even in the presence of wildcards.

This first PR will ruin optimization because we haven't updated what
"wildcard" means but we've labeled more stuff as wildcards. That's the
only safe way to do it, but I'll stack something on top to fix it.

Differential Revision: [D15447568](https://our.internmc.facebook.com/intern/diff/D15447568)

suo committed:
[jit] Mark values entering containers as wildcards

This is the first step to improving wildcard analysis.

The end-state is that the wildcard set now represents everything we're
not sure about. Inputs and values that enter/exit containers will be
part of the wildcard set. There will be one wildcard set per type.

The key difference is that today, a wildcard value may alias *any other
value* in the graph. After this change, the wildcard value may alias
*any other wildcard*, but will not alias other stuff.

So if we have a lot of purely functional code in an inner block, we can
optimize it even in the presence of wildcards.

This first PR will ruin optimization because we haven't updated what
"wildcard" means but we've labeled more stuff as wildcards. That's the
only safe way to do it, but I'll stack something on top to fix it.

suo committed:
Update on "[jit] Mark values entering containers as wildcards"

[jit] Mark values entering containers as wildcards

This is the first step to improving wildcard analysis.

The end-state is that the wildcard set now represents everything we're
not sure about. Inputs and values that enter/exit containers will be
part of the wildcard set. There will be one wildcard set per type.

The key difference is that today, a wildcard value may alias *any other
value* in the graph. After this change, the wildcard value may alias
*any other wildcard*, but will not alias other stuff.

So if we have a lot of purely functional code in an inner block, we can
optimize it even in the presence of wildcards.

This first PR will ruin optimization because we haven't updated what
"wildcard" means but we've labeled more stuff as wildcards. That's the
only safe way to do it, but I'll stack something on top to fix it.

gh-metadata: pytorch pytorch 20556 gh/suo/39/head

suo committed:
Update on "[jit] Mark values entering containers as wildcards"

[jit] Mark values entering containers as wildcards

This is the first step to improving wildcard analysis.

The end-state is that the wildcard set now represents everything we're
not sure about. Inputs and values that enter/exit containers will be
part of the wildcard set. There will be one wildcard set per type.

The key difference is that today, a wildcard value may alias *any other
value* in the graph. After this change, the wildcard value may alias
*any other wildcard*, but will not alias other stuff.

So if we have a lot of purely functional code in an inner block, we can
optimize it even in the presence of wildcards.

This first PR will ruin optimization because we haven't updated what
"wildcard" means but we've labeled more stuff as wildcards. That's the
only safe way to do it, but I'll stack something on top to fix it.

gh-metadata: pytorch pytorch 20556 gh/suo/39/head

eellison commented:
I know this will be landed in a stack so you may already account for these things, but a few thoughts. 

- The way we are marking wildcards currently is by adding value `v` to a wildcard set. In the future we will need to add all values that may alias `v` to the set as well. 
- If there is e.g. a graph that takes in a Tensor and a List of Tensor, the tensor needs to be in the wildcard set, because it could be contained in the list.
- eventually it would be nice to have different wildcard sets per types or some other way of making it so that two wildcard values of different types don't alias each other.

suo commented:
@eellison I've just stacked a PR on top of this one that should address your comments.

suo committed:
Update on "[jit] Mark values entering containers as wildcards"

[jit] Mark values entering containers as wildcards

This is the first step to improving wildcard analysis.

The end-state is that the wildcard set now represents everything we're
not sure about. Inputs and values that enter/exit containers will be
part of the wildcard set. There will be one wildcard set per type.

The key difference is that today, a wildcard value may alias *any other
value* in the graph. After this change, the wildcard value may alias
*any other wildcard*, but will not alias other stuff.

So if we have a lot of purely functional code in an inner block, we can
optimize it even in the presence of wildcards.

This first PR will ruin optimization because we haven't updated what
"wildcard" means but we've labeled more stuff as wildcards. That's the
only safe way to do it, but I'll stack something on top to fix it.

gh-metadata: pytorch pytorch 20556 gh/suo/39/head

suo committed:
Update on "[jit] Mark values entering containers as wildcards"

[jit] Mark values entering containers as wildcards

This is the first step to improving wildcard analysis.

The end-state is that the wildcard set now represents everything we're
not sure about. Inputs and values that enter/exit containers will be
part of the wildcard set. There will be one wildcard set per type.

The key difference is that today, a wildcard value may alias *any other
value* in the graph. After this change, the wildcard value may alias
*any other wildcard*, but will not alias other stuff.

So if we have a lot of purely functional code in an inner block, we can
optimize it even in the presence of wildcards.

This first PR will ruin optimization because we haven't updated what
"wildcard" means but we've labeled more stuff as wildcards. That's the
only safe way to do it, but I'll stack something on top to fix it.

gh-metadata: pytorch pytorch 20556 gh/suo/39/head
